app/controllers/autotune/themes_controller.rb

```diff
@@ -0,0 +1,120 @@
+require_dependency 'autotune/application_controller'
+
+module Autotune
+  # API for themes
+  class ThemesController < ApplicationController
+    before_action :respond_to_html
+    model Theme
+
+    rescue_from ActiveRecord::UnknownAttributeError do |exc|
+      render_error exc.message, :bad_request
+    end
+
+    before_action :only => [:show, :create, :update, :destroy] do
+      unless current_user.role?(:superuser) ||
+             current_user.role?(:designer => instance.group.name)
+        render_error 'Forbidden', :forbidden
+      end
+    end
+
+    before_action :only => [:new, :create] do
+      unless current_user.role? [:superuser, :designer]
+        render_error 'Forbidden', :forbidden
+      end
+    end
+
+    def new; end
+
+    def edit; end
+
+    def index
+      @themes = current_user.designer_themes
+
+      # Filter and search query
+      query = {}
+
+      query[:status] = params[:status] if params.key? :status
+      query[:group_id] = params[:group] if params.key? :group
+
+      @themes = @themes.search(params[:search], :title) if params.key? :search
+
+      @themes = @themes.where(query)
+
+      page = params[:page] || 1
+      per_page = params[:per_page] || 15
+      @themes = @themes.paginate(:page => page, :per_page => per_page)
+      link_str = '<%s>; rel="%s"'
+      links = [
+        link_str % [
+          themes_url(:page => @themes.current_page, :per_page => per_page), 'page'],
+        link_str % [
+          themes_url(:page => 1, :per_page => per_page), 'first'],
+        link_str % [
+          themes_url(:page => @themes.total_pages, :per_page => per_page), 'last']
+      ]
+      if @themes.next_page
+        links << link_str % [
+          themes_url(:page => @themes.next_page, :per_page => per_page), 'next']
+      end
+      if @themes.previous_page
+        links << link_str % [
+          themes_url(:page => @themes.previous_page, :per_page => per_page), 'prev']
+      end
+      headers['Link'] = links.join(', ')
+      headers['X-Total'] = @themes.count
+    end
+
+    def show
+      @theme = instance
+    end
+
+    def create
+      @theme = Theme.new
+      @theme.attributes = select_from_post :title, :data, :group_id, :slug
+      @theme.parent = Theme.get_default_theme_for_group(@theme.group_id)
```

Comment:
Could this be part of validation? like how the sluggable helper creates a slug if none exist?